nomad/host_volume_endpoint_test.go

```diff
@@ -0,0 +1,571 @@
+				Namespace: ns,
+				AuthToken: token},
+		}
+
+		c1.setCreate(nil, "should not call this endpoint on register RPC")
+
+		ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+		t.Cleanup(cancel)
+		volCh := make(chan *structs.HostVolume)
+		errCh := make(chan error)
+
+		getReq := &structs.HostVolumeGetRequest{
+			ID: vol1ID,
+			QueryOptions: structs.QueryOptions{
+				Region:        srv.Region(),
+				Namespace:     ns,
+				AuthToken:     token,
+				MinQueryIndex: expectIndex,
+			},
+		}
+
+		go func() {
+			codec := rpcClient(t, srv)
+			var getResp structs.HostVolumeGetResponse
+			err := msgpackrpc.CallWithCodec(codec, "HostVolume.Get", getReq, &getResp)
+			if err != nil {
+				errCh <- err
+			}
+			volCh <- getResp.Volume
+		}()
+
+		var registerResp structs.HostVolumeRegisterResponse
+		err = msgpackrpc.CallWithCodec(codec, "HostVolume.Register", registerReq, &registerResp)
+		must.NoError(t, err)
+
+		select {
+		case <-ctx.Done():
+			t.Fatal("timeout or cancelled")
+		case vol := <-volCh:
+			must.Greater(t, expectIndex, vol.ModifyIndex)
+		case err := <-errCh:
+			t.Fatalf("unexpected error: %v", err)
+		}
+	})
+
+	t.Run("delete blocked by allocation claims", func(t *testing.T) {
+		vol2, err := store.HostVolumeByID(nil, ns, vol2ID, false)
+		must.NoError(t, err)
+		must.NotNil(t, vol2)
+
+		// claim one of the volumes with a pending allocation
+		alloc := mock.MinAlloc()
+		alloc.NodeID = node1.ID
+		alloc.Job.TaskGroups[0].Volumes = map[string]*structs.VolumeRequest{"example": {
+			Name:   "example",
+			Type:   structs.VolumeTypeHost,
+			Source: vol2.Name,
+		}}
+		index++
+		must.NoError(t, store.UpsertAllocs(structs.MsgTypeTestSetup,
+			index, []*structs.Allocation{alloc}))
+
+		delReq := &structs.HostVolumeDeleteRequest{
+			VolumeIDs: []string{vol1ID, vol2ID},
+			WriteRequest: structs.WriteRequest{
+				Region:    srv.Region(),
+				Namespace: ns,
+				AuthToken: token},
+		}
+		var delResp structs.HostVolumeDeleteResponse
+
+		err = msgpackrpc.CallWithCodec(codec, "HostVolume.Delete", delReq, &delResp)
+		must.EqError(t, err, "Permission denied")
+
+		delReq.AuthToken = powerToken
+		err = msgpackrpc.CallWithCodec(codec, "HostVolume.Delete", delReq, &delResp)
+		must.EqError(t, err, fmt.Sprintf("volume %s in use by allocations: [%s]", vol2ID, alloc.ID))
+
+		// volume not in use will be deleted even if we get an error
+		getReq := &structs.HostVolumeGetRequest{
+			ID: vol1ID,
+			QueryOptions: structs.QueryOptions{
+				Region:    srv.Region(),
+				Namespace: ns,
+				AuthToken: token,
+			},
+		}
+		var getResp structs.HostVolumeGetResponse
+		err = msgpackrpc.CallWithCodec(codec, "HostVolume.Get", getReq, &getResp)
+		must.NoError(t, err)
+		must.Nil(t, getResp.Volume)
+
+		// update the allocations terminal so the delete works
+		alloc = alloc.Copy()
+		alloc.ClientStatus = structs.AllocClientStatusFailed
+		nArgs := &structs.AllocUpdateRequest{
+			Alloc: []*structs.Allocation{alloc},
```

Comment:
`Alloc` field is marked as deprecated :shrug:

Reply:
The client does it the deprecated way still, with a compatibility note on the struct for Nomad 0.11 :laughing: So I'm going to leave this as-is so we're testing the behavior that actually exists.